File: src-code/scripts/expr.py
#!/usr/bin/env python3

from io import TextIOWrapper

FILE_NAME = "src/main/java/jh/craft/interpreter/representation/Expr.java"

CLASSES = { 
    'Binary'   : 'Expr left, Token operator, Expr right',
    'Literal'  : 'Object value',
    'Grouping' : 'Expr expression',
    'Unary'    : 'Token operator, Expr expression'
}

def join_lines(lines):
    return '\n'.join(lines)


def generate_classes( file : TextIOWrapper ) :
    file.write(f"""
package jh.craft.interpreter.representation;

import jh.craft.interpreter.scanner.Token;

public interface Expr {{

    interface Visitor<T> {{
{ join_lines(
    [
        f'        T visit{cname}( {cname} {cname.lower()} );' 
        for cname in CLASSES
    ]
)}
    }}

    <T> T accept( Visitor<T> visitor );
{
    join_lines([
 f'''
    record {cname}( {args} ) implements Expr {{
        @Override
        public <T> T accept( Visitor<T> visitor ){{ 
            return visitor.visit{cname}( this );
        }}
    }}'''   for cname, args in CLASSES.items()
    ])
}

}}""")
   
def main():
    with open(FILE_NAME, "w") as file:
        generate_classes(file)

if __name__ == "__main__":
    main()

File: src-code/src/main/java/jh/craft/interpreter/representation/Expr.java

package jh.craft.interpreter.representation;

import jh.craft.interpreter.scanner.Token;
import java.util.List;

public interface Expr {

    interface Visitor<T> {
        T visitBinary( Binary binary );
        T visitLiteral( Literal literal );
        T visitGrouping( Grouping grouping );
        T visitUnary( Unary unary );
        T visitVariable( Variable variable );
        T visitAssign( Assign assign );
        T visitLogical( Logical logical );
        T visitCall( Call call );
    }

    <T> T accept( Visitor<T> visitor );

    record Binary( Expr left, Token operator, Expr right ) implements Expr {
        @Override
        public <T> T accept( Visitor<T> visitor ){ 
            return visitor.visitBinary( this );
        }
    }

    record Literal( Object value ) implements Expr {
        @Override
        public <T> T accept( Visitor<T> visitor ){ 
            return visitor.visitLiteral( this );
        }
    }

    record Grouping( Expr expression ) implements Expr {
        @Override
        public <T> T accept( Visitor<T> visitor ){ 
            return visitor.visitGrouping( this );
        }
    }

    record Unary( Token operator, Expr expression ) implements Expr {
        @Override
        public <T> T accept( Visitor<T> visitor ){ 
            return visitor.visitUnary( this );
        }
    }

    record Variable( Token name ) implements Expr {
        @Override
        public <T> T accept( Visitor<T> visitor ){ 
            return visitor.visitVariable( this );
        }
    }

    record Assign( Token name, Expr value ) implements Expr {
        @Override
        public <T> T accept( Visitor<T> visitor ){ 
            return visitor.visitAssign( this );
        }
    }

    record Logical( Expr left, Token operator, Expr right ) implements Expr {
        @Override
        public <T> T accept( Visitor<T> visitor ){ 
            return visitor.visitLogical( this );
        }
    }

    record Call( Expr callee, Token rightParen, List<Expr> arguments ) implements Expr {
        @Override
        public <T> T accept( Visitor<T> visitor ){ 
            return visitor.visitCall( this );
        }
    }

}